### ISSUE TITLE
Shaded Jackson

### ISSUE THREAD

**Issue by smithagrg** [deeplearning4j/deeplearning4j#8437]
We are implementing a neural network for one of our products using the deep learning libraries  for java. One of the dependencies for deeplearning4j-nn.jar is shaded Jackson by nd4j. This is blowing up in the OWASP security scan of the libraries. Therefore, we cannot use this dependency. Is there an alternate option around the shaded jar?

Below are  all the jars we are using for this project. This is the smallest list of dependencies we could find to get the model running.  I would like to know if we can reduce this further down and/or reduce the size of some of the big ones.? Please guide me.

#### Version Information
deeplearning4j-nn-1.0.0-beta5.jar;
guava-1.0.0-beta5.jar;
jackson-1.0.0-beta5.jar;
javacpp-1.5.1-1.jar;
mkl-dnn-0.20-1.5.1-linux-x86_64.jar;
mkl-dnn-0.20-1.5.1-windows-x86_64.jar;
nd4j-api-1.0.0-beta5.jar;
nd4j-buffer-1.0.0-beta5.jar;
nd4j-common-1.0.0-beta5.jar;
nd4j-context-1.0.0-beta5.jar;
nd4j-native-1.0.0-beta5.jar;
nd4j-native-1.0.0-beta5-linux-x86_64.jar;
nd4j-native-1.0.0-beta5-windows-x86_64.jar;
nd4j-native-api-1.0.0-beta5.jar;
nd4j-native-platform-1.0.0-beta5.jar;
openblas-0.3.6-1.5.1.jar;
openblas-0.3.6-1.5.1-linux-x86_64.jar;
openblas-0.3.6-1.5.1-windows-x86_64.jar;

**Comment by AlexDBlack** [comment]
> One of the dependencies for deeplearning4j-nn.jar is shaded Jackson by nd4j. This is blowing up in the OWASP security scan of the libraries.
> Is there an alternate option around the shaded jar?

Internally it's Jackson 2.9.9 and 2.9.9.3
https://github.com/eclipse/deeplearning4j/blob/master/pom.xml#L331-L332

That was, at the time of upgrade, the latest version. It's possible more vulnerabilities have been discovered since.

There is no way to remove this library, it's integral to everything we do with JSON (including reading and writing model formats).

FWIW we use a shaded version to avoid dependency conflicts, without it users are pretty much guaranteed to run into frequent dependency conflicts with common libraries such as Apache Spark and web frameworks. For example Spark 2.4.4 uses jackson 2.6.7 from 2016.
So not shading Jackson isn't an option.

We can upgrade to the latest (2.10.1, 10 November 2019, from 2.9.9 16 May 2019) - would that help?
Is the OWASP scanning complaining about the underlying Jackson version, or the fact that it's shaded?

As for dependencies - if you only need one operating system for deployments, use "nd4j-native" instead of nd4j-native-platform.
See also https://github.com/bytedeco/javacpp-presets/wiki/Reducing-the-Number-of-Dependencies

**Comment by smithagrg** [comment]
Hi Alex,

 The issue is in fact with the Jackson version. Below is the error from
OWASP scan. So, upgrading would definitely help.

 "*A Polymorphic Typing issue was discovered in FasterXML jackson-databind
before 2.9.10. It is related to com.zaxxer.hikari.HikariConfig*."

*Dependencies*

We deploy in linux  or windows platforms. Our maven configuration is
limited and forces us to pretty much enter every dependency separately. If
I don't use the "native-platform", are you saying that we don't need to add
any platform specific dependencies like the windows and linux specific ones?

Thanks

Smitha

**Comment by AlexDBlack** [comment]
> So, upgrading would definitely help.

Thanks, I'll look at upgrading to 2.10.1

> We deploy in linux  or windows platforms. Our maven configuration is
limited and forces us to pretty much enter every dependency separately. If
I don't use the "native-platform", are you saying that we don't need to add
any platform specific dependencies like the windows and linux specific ones?

The way it works:
nd4j-native-platform (or cuda-x.x-platform) includes native binaries for all OSs.
if you use nd4j-native instead, it includes it only for the _build_ platform. So linux binaries only if you build on linux, windows only if built on windows, etc.
That works fine until you build on windows and deploy on linux, for example (no linux binaries).

So:
* For same build/deploy OS - just use nd4j-native
* For different build/deploy OS - or multiple deployment OSs in the one JAR - you can use nd4j-native  dependency, _and_ nd4j-native dependency with classifiers. That's basically equivalent to nd4j-native-platform + exclusions though, in practice

You can see a list of classifiers - and example of classifier dependencies - here: https://deeplearning4j.org/cpu

**Comment by AlexDBlack** [comment]
Jackson has been upgraded to 2.10.1 (latest) here: https://github.com/KonduitAI/deeplearning4j/pull/82
That will be merged soon, and will be in the next release.

I'll close this as resolved.

**Comment by smithagrg** [comment]
Thank you, when will the next release be on?

Smitha

**Comment by smithagrg** [comment]
Our maven plugin does not automatically download dependencies, so I
download the linux and windows dependency using the classifier.  It is
fine. Is there a smaller alternative for the mkl-dnn and open-blas
platform libraries?
I am trying to understand why we have platform libraries for those as well?
For example, for a windows environment I seem to need the following minimum
set of platform dependencies.  Is my understanding right?
Is there a smaller alternative for the mkl-dnn platform dependency for
example, because that is a huge jar.

mkl-dnn-0.20-1.5.1-windows-x86_64.jar
nd4j-native-windows-x86_64.jar
openblas-0.3.6-1.5.1-windows-x86_64.jar

Smitha

**Comment by raver119** [comment]
mkl-dnn shouldn't be needed in beta5 and subsequent releases, since it's compiled into libnd4j staticallym and comes in nd4j-native-windows-x86_64.jar. OpenBLAS or MKL is required though.

**Comment by saudet** [comment]
@smithagrg We can use profiles to manage the native dependencies:
https://github.com/bytedeco/javacpp-presets/wiki/Reducing-the-Number-of-Dependencies

**Comment by smithagrg** [comment]
I get an error without it

Exception in thread "pool-2-thread-1" java.lang.ExceptionInInitializerError
Caused by: java.lang.UnsatisfiedLinkError: no jnind4jcpu in
java.library.path
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1867)
at java.lang.Runtime.loadLibrary0(Runtime.java:870)
at java.lang.System.loadLibrary(System.java:1122)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1487)
at org.bytedeco.javacpp.Loader.load(Loader.java:1190)
at org.bytedeco.javacpp.Loader.load(Loader.java:1040)
at org.nd4j.nativeblas.Nd4jCpu.<clinit>(Nd4jCpu.java:10)
at java.lang.Class.forName0(Native Method)
at java.lang.Class.forName(Class.java:264)
at org.nd4j.nativeblas.NativeOpsHolder.<init>(NativeOpsHolder.java:84)
... 16 more
Caused by: java.lang.UnsatisfiedLinkError:
C:\Users\svarghese\.javacpp\cache\nd4j-native-windows-x86_64.jar\org\nd4j\nativeblas\windows-x86_64\jnind4jcpu.dll:
Can't find dependent libraries
at java.lang.ClassLoader$NativeLibrary.load(Native Method)
at java.lang.ClassLoader.loadLibrary0(ClassLoader.java:1941)
at java.lang.ClassLoader.loadLibrary(ClassLoader.java:1824)
at java.lang.Runtime.load0(Runtime.java:809)
at java.lang.System.load(System.java:1086)
at org.bytedeco.javacpp.Loader.loadLibrary(Loader.java:1461)
... 22 more

**Comment by smithagrg** [comment]
Do you know when the next release is ? If not, can you send me the pom file to create a shaded jar myself with the versions we require?

Thanks

> > One of the dependencies for deeplearning4j-nn.jar is shaded Jackson by nd4j. This is blowing up in the OWASP security scan of the libraries.
> > Is there an alternate option around the shaded jar?
> 
> Internally it's Jackson 2.9.9 and 2.9.9.3
> https://github.com/eclipse/deeplearning4j/blob/master/pom.xml#L331-L332
> 
> That was, at the time of upgrade, the latest version. It's possible more vulnerabilities have been discovered since.
> 
> There is no way to remove this library, it's integral to everything we do with JSON (including reading and writing model formats).
> 
> FWIW we use a shaded version to avoid dependency conflicts, without it users are pretty much guaranteed to run into frequent dependency conflicts with common libraries such as Apache Spark and web frameworks. For example Spark 2.4.4 uses jackson 2.6.7 from 2016.
> So not shading Jackson isn't an option.
> 
> We can upgrade to the latest (2.10.1, 10 November 2019, from 2.9.9 16 May 2019) - would that help?
> Is the OWASP scanning complaining about the underlying Jackson version, or the fact that it's shaded?
> 
> As for dependencies - if you only need one operating system for deployments, use "nd4j-native" instead of nd4j-native-platform.
> See also https://github.com/bytedeco/javacpp-presets/wiki/Reducing-the-Number-of-Dependencies

**Comment by raver119** [comment]
We are building release binaries right now. So expect it +- tomorrow.

**Comment by smithagrg** [comment]
Thank you....we are experiencing another issue with the shaded jackson with
the system properties being set in the META-INF/Services. It is altering
behavior in other parts of our application which uses a different version
of Jackson as well. This means that I am not able to use the maven version
of the jar as is without modifying. We are also concerned about other
impacts of this dependency. Do you have any suggestions? What is the need
to set these properties? The neural network is training and predicting
without any errors once I remove the settings.

Smitha

**Comment by AlexDBlack** [comment]
@smithagrg META-INF/services is for java service loader mechanism.
I took a look there, most of those are shaded (and hence shouldn't cause problems) except for javax.xml.stream.XMLOutputFactory, .XMLInputFactory and .XMLEventFactory
Are these what you are referring to?

They aren't added manually by us, but (as far as I can tell) simply part of jackson-dataformat-xml that is carried through the shading process.
If they are the problem, we can probably exclude them, though I'd need to look into that further (and test of course).

If you are building an uber-jar, it should be possible to exclude those manually easily enough. However, doing it in a way that doesn't _also_ exclude the service loader file version (of same name) from the other Jackson version you want to keep is more difficult. Not sure of a good solution for that yet TBH.

**Comment by smithagrg** [comment]
Yeah...they are what is causing issues.

Smitha

**Comment by smithagrg** [comment]
We are not building an uber.jar. Infact we are adding each dependency required in pom.  But, I think it would be better for us to create this shaded jackson jar ourselves so we can control the version and such. I am not savvy with maven and am not sure how to get that accomplished.  Can you maybe send me a pom version to get this shaded jackson with relocation to "nd4j" package?

**Comment by AlexDBlack** [comment]
The shading occurs here: https://github.com/eclipse/deeplearning4j/blob/master/nd4j/nd4j-shade/jackson/pom.xml
In theory it's as simple as excluding those `META-INF/services/javax.xml.stream` files using filters like this:
https://stackoverflow.com/a/46046022

Unfortunately the release got delayed (problems with our cloud provider), upside is I may be able to find time to look at this for 1.0.0-beta6...

**Comment by smithagrg** [comment]
Thanks Alex for the help. I managed to create the shaded jar today and it
seem to work. We are just a bit concerned right now about more potential
issues that could be caused by the shaded jars.  Also, you mentioned the
next release to be beta 6? When are we planning to get out of beta? Do you
think it would be safe to use the beta version in an enterprise level
software?

As for the shaded Jackson version, do you need the Woodstox package? What
is it used for?

Thanks again for your prompt responses

Smitha

**Comment by AlexDBlack** [comment]
> We are just a bit concerned right now about more potential issues that could be caused by the shaded jars.

Shaded jars are a tradeoff - and the right one IMO.
The reason we use them is to avoid dependency issues. For example the Jackson libraries are widely used, and users are pretty much guaranteed to run into dependency problems any time they run DL4J or ND4J with Apache Spark, any of the web frameworks (Play, Vertx, etc), or with a ton of other popular libraries. In short, not shading isn't an option.
Downgrading version isn't an option either - if we match the version for Spark, we still don't match it for say Vertx. That's not to mention the vulnerabilities with earlier versions (like what Spark uses).
OSGi is another option (allows multiple versions of the one library) but that's not suitable for 95% of the userbase.

We've been using shaded jars for years, and (this issue aside) it's been pretty reliable - far more so than any of the alternatives.


> Also, you mentioned the next release to be beta 6? When are we planning to get out of beta? Do you think it would be safe to use the beta version in an enterprise level software?

The only reason it's "beta" is that the API isn't completely stable yet, and there's a few final features to add before we can consider the library essentially feature complete.
Next release (end of Q1 2020) is when we are targeting the full 1.0.0 release. (It had been planned earlier, but we're not there yet)
It's totally fine for enterprise, we have enterprise customers for example 1.0.0-beta5 in production already.

> As for the shaded Jackson version, do you need the Woodstox package? What is it used for?

IIRC it's a transient dependency used internally in Jackson. Different versions of it are also a cause for dependency problems (NoClassDefFound etc exceptions at runtime)

**Comment by AlexDBlack** [comment]
@smithagrg I've added that exclusion here:
https://github.com/eclipse/deeplearning4j/pull/8529

nd4j-jackson service loader (META-INF/services) now looks like this:
![image](https://user-images.githubusercontent.com/2360237/70796355-264d9380-1df6-11ea-98b7-074f3e3521f4.png)

So just those 3 javax.xml.stream.XMLOutputFactory, .XMLInputFactory and .XMLEventFactory files have been removed; remaiving service loader files are for our shaded libraries only.
I tested this as well, looks like there's no side effects on our end.

This will make it into the 1.0.0-beta6 release.

**Comment by smithagrg** [comment]
Thankyou very much Alex. When will be the beta release be again?

Smitha

**Comment by AlexDBlack** [comment]
Assuming we don't run into any more issues with our cloud provider, probably within 24 hours.
Early next week at the latest.

**Comment by smithagrg** [comment]
Many thanks Alex. We got the new build and all our current issues are
resolved.

Smitha

**Comment by tanjucat** [comment]
@AlexDBlack this is an issue again now in 2025. We are blocked by vulnerability findings in some packages (like jackson) inside nd4j. Is it possible to have a version without the shaded libraries so that they can be overriden when the security updates are available?

**Comment by agibsonccc** [comment]
@tanjucat  alex isn't with the project anymore. Please open new issues if you have something specific.  Just override them using excludes as you need. I'm happy to discuss this as needed. If you want custom releases you're welcome to pay for them as well. Given the resources of the project mainly being focused on internal c++ modernization right now the only upgrades going out based on old releases are paid work. Please email support@konduit.ai if you would like to arrange something.
